codes/typescript/chapter_dynamic_programming/climbing_stairs_constraint_dp.ts

```diff
@@ -0,0 +1,35 @@
+/**
+ * File: climbing_stairs_constraint_dp.ts
+ * Created Time: 2023-08-23
+ * Author: Gaofer Chou (gaofer-chou@qq.com)
+ */
+
+/* 带约束爬楼梯：动态规划 */
+function climbingStairsConstraintDP(n: number): number {
+    if (n == 1 || n == 2) {
+        return n;
+    }
+    // 初始化 dp 表，用于存储子问题的解
+    let dp: Array<Array<number>> = Array.from(
+        new Array(n + 1),
+        () => new Array(3)
+    );
+    // 初始状态：预设最小子问题的解
+    dp[1][1] = 1;
+    dp[1][2] = 0;
+    dp[2][1] = 0;
+    dp[2][2] = 1;
+    // 状态转移：从较小子问题逐步求解较大子问题
+    for (let i: number = 3; i <= n; i++) {
+        dp[i][1] = dp[i - 1][2];
+        dp[i][2] = dp[i - 2][1] + dp[i - 2][2];
+    }
+    return dp[n][1] + dp[n][2];
+}
+
+/* Driver Code */
+const n: number = 9;
+let res: number = climbingStairsConstraintDP(n);

```

Comment:
```suggestion
const res = climbingStairsConstraintDP(n);
```

Reply:
Because Typescript has Type Inference, I think this variable is so clear and the function has defined the return type, so do not repeat yourself.  I recommend you give the function type, which can control our input and output, and like the document. Second, give the data structure type. 
Please check others and make your code clean.